rosetta/services/operation_components.go

```diff
@@ -0,0 +1,232 @@
+package services
+
+import (
+	"fmt"
+	"math/big"
+
+	"github.com/coinbase/rosetta-sdk-go/types"
+
+	"github.com/harmony-one/harmony/rosetta/common"
+	"github.com/pkg/errors"
+)
+
+const (
+	// maxNumOfConstructionOps ..
+	maxNumOfConstructionOps = 2
+)
+
+// OperationComponents are components from a set of operations to construct a valid transaction
+type OperationComponents struct {
+	Type           string                   `json:"type"`
+	From           *types.AccountIdentifier `json:"from"`
+	To             *types.AccountIdentifier `json:"to"`
+	Amount         *big.Int                 `json:"amount"`
+	StakingMessage interface{}              `json:"staking_message,omitempty"`
+}
+
+// IsStaking ..
+func (s *OperationComponents) IsStaking() bool {
+	return s.StakingMessage != nil
+}
+
+// GetOperationComponents ensures the provided operations creates a valid transaction and returns
+// the OperationComponents of the resulting transaction.
+//
+// Providing a gas expenditure operation is INVALID.
+// All staking & cross-shard operations require metadata matching the operation type to be a valid.
+// All other operations do not require metadata.
+// TODO (dm): implement staking transaction construction
+func GetOperationComponents(
+	operations []*types.Operation,
+) (*OperationComponents, *types.Error) {
+	if operations == nil {
+		return nil, common.NewError(common.CatchAllError, map[string]interface{}{
+			"message": "nil operations",
+		})
+	}
+	if len(operations) > maxNumOfConstructionOps || len(operations) == 0 {
+		return nil, common.NewError(common.InvalidTransactionConstructionError, map[string]interface{}{
+			"message": fmt.Sprintf("invalid number of operations, must <= %v & > 0", maxNumOfConstructionOps),
+		})
+	}
+
+	if len(operations) == 2 {
```

Comment:
== maxNumOfConstructionOps ?

Reply:
`maxNumberOfConstructionOps` can change in the future. I'll just make a const for it. This should also take care of the comment for getTransferOperationComponents